src/event/device.rs

```diff
@@ -0,0 +1,303 @@
+//! Raw hardware events that are not associated with any particular window.
+//!
+//! Useful for interactions that diverge significantly from a conventional 2D GUI, such as 3D camera or first-person
+//! game controls. Many physical actions, such as mouse movement, can produce both device and window events. Because
+//! window events typically arise from virtual devices (corresponding to GUI cursors and keyboard focus) the device IDs
+//! may not match.
+//!
+//! All attached devices are guaranteed to emit an `Added` event upon the initialization of the event loop.
+//!
+//! Note that device events are always delivered regardless of window focus.
+
+use platform_impl;
+use event::{AxisId, ButtonId, ElementState, KeyboardInput, MouseButton};
+use event_loop::EventLoop;
+use std::{fmt, io};
+
+/// A hint suggesting the type of button that was pressed.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum GamepadButton {
+    Start,
+    Select,
+
+    /// The north face button.
+    ///
+    /// * Nintendo: X
+    /// * Playstation: Triangle
+    /// * XBox: Y
+    North,
+    /// The south face button.
+    ///
+    /// * Nintendo: B
+    /// * Playstation: X
+    /// * XBox: A
+    South,
+    /// The east face button.
+    ///
+    /// * Nintendo: A
+    /// * Playstation: Circle
+    /// * XBox: B
+    East,
+    /// The west face button.
+    ///
+    /// * Nintendo: Y
+    /// * Playstation: Square
+    /// * XBox: X
+    West,
+
+    LeftStick,
+    RightStick,
+
+    LeftTrigger,
+    RightTrigger,
+
+    LeftShoulder,
+    RightShoulder,
+
+    DPadUp,
+    DPadDown,
+    DPadLeft,
+    DPadRight,
+}
+
+/// A hint suggesting the type of axis that moved.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum GamepadAxis {
+    LeftStickX,
+    LeftStickY,
+
+    RightStickX,
+    RightStickY,
+
+    LeftTrigger,
+    RightTrigger,
+
+    // @francesca64 review: why were these variants here? I don't see how it makes sense for the dpad or hat switch
+    // to have axes, since they're both four separate buttons.
+
+    // /// This is supposed to have a specialized meaning, referring to a point-of-view switch present on joysticks used
+    // /// for flight simulation. However, Xbox 360 controllers (and their derivatives) use a hat switch for the D-pad.
+    // HatSwitch,
+    // DPadUp,
+    // DPadDown,
+    // DPadLeft,
+    // DPadRight,
```

Comment:
Copied from comment:
@francesca64 review: why were these variants here? I don't see how it makes sense for the dpad or hat switch to have axes, since they're both four separate buttons.

Reply:
I don't remember, but if there's any validity to them being here, it probably has to do with weird differences in how Xbox controllers work under RAWINPUT or something. There are lots of articles that complain about quirks like that, so if nothing comes up from a cursory search, then you can just assume this is some cruft I left in.